clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -121,29 +375,174 @@ mlir::Value CIRGenFunction::emitScalarExpr(const Expr *e) {
   return ScalarExprEmitter(*this, builder).Visit(const_cast<Expr *>(e));
 }
 
+[[maybe_unused]] static bool MustVisitNullValue(const Expr *e) {
+  // If a null pointer expression's type is the C++0x nullptr_t, then
+  // it's not necessarily a simple constant and it must be evaluated
+  // for its potential side effects.
+  return e->getType()->isNullPtrType();
+}
+
 // Emit code for an explicit or implicit cast.  Implicit
 // casts have to handle a more broad range of conversions than explicit
 // casts, as they handle things like function to ptr-to-function decay
 // etc.
 mlir::Value ScalarExprEmitter::VisitCastExpr(CastExpr *ce) {
-  Expr *e = ce->getSubExpr();
+  Expr *subExpr = ce->getSubExpr();
   QualType destTy = ce->getType();
   CastKind kind = ce->getCastKind();
 
+  // These cases are generally not written to ignore the result of evaluating
+  // their sub-expressions, so we clear this now.
+  ignoreResultAssign = false;
```

Comment:
Is this variable used?

Reply:
See my comment here: https://github.com/llvm/llvm-project/pull/130690#discussion_r1990214545

At the moment no, but I think it would be more error prone to add it later than to add it as a dead variable now.

Reply:
I agree.

Reply:
You'll have to do a void cast on it then, else this will cause a warning.

Reply:
I spent a while trying to repro the warning I could swear I've seen before, but I've not been able to get this to happen.  So disregard.